File: src/main/java/me/ruslanys/vkmusic/config/ExecutorsConfig.java
package me.ruslanys.vkmusic.config;

import org.slf4j.LoggerFactory;
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.annotation.AsyncConfigurer;
import org.springframework.scheduling.annotation.EnableAsync;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.scheduling.annotation.SchedulingConfigurer;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
import org.springframework.scheduling.config.ScheduledTaskRegistrar;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

/**
 * Date: 02.02.16
 * Time: 14:40
 *
 * @author Ruslan Molchanov (ruslanys@gmail.com)
 * @author http://mruslan.com
 */
@Configuration
@EnableScheduling
@EnableAsync
public class ExecutorsConfig implements SchedulingConfigurer, AsyncConfigurer {

    @Override
    public void configureTasks(ScheduledTaskRegistrar taskRegistrar) {
        ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();
        taskScheduler.setPoolSize(5);
        taskScheduler.initialize();
        taskScheduler.setThreadNamePrefix("ScheduledExecutor-");

        taskRegistrar.setTaskScheduler(taskScheduler);
    }

    @Override
    public Executor getAsyncExecutor() {
        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
        executor.setCorePoolSize(5);
        executor.setThreadNamePrefix("AsyncExecutor-");
        executor.initialize();
        return executor;
    }

    @Override
    public AsyncUncaughtExceptionHandler getAsyncUncaughtExceptionHandler() {
        return (throwable, method, objects) -> LoggerFactory.getLogger("Async").error("Async error", throwable); // ?
    }

    @Bean
    public ScheduledExecutorService scheduledExecutorService() {
         return Executors.newSingleThreadScheduledExecutor();
    }

}

File: src/main/kotlin/me/ruslanys/vkmusic/config/ExecutorsConfig.kt
package me.ruslanys.vkmusic.config

import me.ruslanys.vkmusic.property.DownloadProperties
import me.ruslanys.vkmusic.service.PropertyService
import org.slf4j.LoggerFactory
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler
import org.springframework.context.ApplicationListener
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.event.ContextStoppedEvent
import org.springframework.scheduling.annotation.AsyncConfigurer
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import java.util.concurrent.Executor

@EnableAsync
@Configuration
class ExecutorsConfig(private val propertyService: PropertyService) : AsyncConfigurer, ApplicationListener<ContextStoppedEvent> {

    @Bean(name = ["asyncExecutor"])
    override fun getAsyncExecutor(): Executor {
        val executor = ThreadPoolTaskExecutor()
        executor.corePoolSize = 3
        executor.setThreadNamePrefix("AsyncExecutor-")
        executor.setAwaitTerminationSeconds(10)
        executor.initialize()

        return executor
    }

    @Bean(name = ["downloadExecutor"])
    fun getDownloadExecutor(): Executor {
        val executor = ThreadPoolTaskExecutor()
        executor.corePoolSize = propertyService.get(DownloadProperties::class.java)!!.poolSize
        executor.setThreadNamePrefix("DownloadExecutor-")
        executor.initialize()

        return executor
    }

    override fun getAsyncUncaughtExceptionHandler(): AsyncUncaughtExceptionHandler =
            AsyncUncaughtExceptionHandler { throwable, _, _ ->
                LoggerFactory.getLogger("Async").error("Async error", throwable)
            }

    override fun onApplicationEvent(event: ContextStoppedEvent) {
        val executors = event.applicationContext.getBeansOfType(ThreadPoolTaskExecutor::class.java)
        for (entry in executors) {
            entry.value.shutdown()
        }
    }

}
